File: backend/app/routers/streak.py
"""Training streak — current + longest, recomputed from active days and
persisted onto the single StreakState row.

An "active day" is a real workout, a logged sport session (football/judo/
padel…), or a 10k-step day — workbook ruling T6a.

GET /api/streak — the current streak snapshot
"""
from datetime import date

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session as DBSession

from app.auth import require_auth
from app.db import get_db
from app.models import Activity, Session, SessionExercise, Set, StepsLog, StreakState, User
from app.routers.settings import get_or_create_settings
from app.schemas import StreakOut
from app.streak import compute_streak

router = APIRouter(prefix="/api/streak", tags=["streak"])


def _workout_dates(db: DBSession, user_id: int) -> set[date]:
    """Calendar dates with at least one completed working set (a real workout)."""
    rows = (
        db.query(Session.started_at)
        .join(SessionExercise, Session.id == SessionExercise.session_id)
        .join(Set, SessionExercise.id == Set.session_exercise_id)
        .filter(Session.user_id == user_id, Set.is_completed == True, Set.is_warmup == False)   # noqa: E712
        .distinct()
        .all()
    )
    return {r.started_at.date() for r in rows}


ACTIVE_STEPS_THRESHOLD = 10_000  # a 10k-step day counts as active (T6a)


def _active_dates(db: DBSession, user_id: int) -> set[date]:
    """Dates that keep the streak alive: a real workout, a logged sport
    session, or a 10k-step day. Sample-seeded rows never count."""
    dates = _workout_dates(db, user_id)
    dates |= {
        r.date
        for r in db.query(Activity.date)
        .filter(Activity.user_id == user_id, Activity.source != "sample").distinct().all()
    }
    dates |= {
        r.date
        for r in db.query(StepsLog.date)
        .filter(
            StepsLog.user_id == user_id,
            StepsLog.steps >= ACTIVE_STEPS_THRESHOLD, StepsLog.source != "sample",
        )
        .all()
    }
    return dates


def streak_snapshot(db: DBSession, user_id: int) -> StreakOut:
    """Recompute the streak, persist the snapshot (longest is monotonic), and
    return it. Shared by GET /api/streak and the dashboard."""
    settings = get_or_create_settings(db, user_id)
    result = compute_streak(_active_dates(db, user_id), settings.streak_rest_gap)

    state = db.query(StreakState).filter(StreakState.user_id == user_id).first()
    if not state:
        state = StreakState(user_id=user_id)
        db.add(state)
    state.current_streak = result.current
    state.longest_streak = max(state.longest_streak or 0, result.longest)
    state.last_workout_date = result.last_workout_date
    db.commit()

    return StreakOut(
        current=result.current,
        longest=state.longest_streak,
        last_workout_date=result.last_workout_date,
        rest_gap=settings.streak_rest_gap,
        alive=result.alive,
        at_risk=result.at_risk,
    )


@router.get("", response_model=StreakOut)
def get_streak(db: DBSession = Depends(get_db), current_user: User = Depends(require_auth)):
    return streak_snapshot(db, current_user.id)

File: frontend/src/local/api/streak.js
/**
 * Training streak, local-first — mirrors backend app/streak.py + routers/streak.py.
 *
 * A workout day is any calendar date with at least one completed working set
 * (completed, non-warm-up). The chain survives up to `rest_gap` rest days and
 * breaks when a larger gap passes. Computed over IndexedDB; no persisted
 * snapshot (each device derives it), so `longest` is the longest chain in
 * history under the current gap.
 */
import { db } from '../db'
import { todayIso } from './util'

const DEFAULT_REST_GAP = 1

// Whole-day difference between two 'YYYY-MM-DD' strings (b - a).
function dayDiff(a, b) {
  return Math.round((Date.parse(b + 'T00:00:00Z') - Date.parse(a + 'T00:00:00Z')) / 86400000)
}

// Working set = completed, non-warm-up, real load — mirrors isWorkingSet /
// the streak router's completed-working-set filter.
async function workoutDates() {
  const days = new Set()
  const sets = await db.set.toArray()
  const workingSeUuids = new Set(
    sets.filter(s => s.is_completed && !s.is_warmup).map(s => s.session_exercise_uuid)
  )
  if (workingSeUuids.size === 0) return days
  const ses = await db.session_exercise.toArray()
  const workingSessionUuids = new Set(
    ses.filter(se => workingSeUuids.has(se.uuid)).map(se => se.session_uuid)
  )
  for (const s of await db.session.toArray()) {
    if (workingSessionUuids.has(s.uuid)) days.add(s.started_at.slice(0, 10))
  }
  return days
}

// Pure port of app/streak.py compute_streak.
export function computeStreak(dates, restGap = DEFAULT_REST_GAP, today = null) {
  const t = today || todayIso()
  const days = [...new Set(dates)].sort()
  if (days.length === 0) {
    return { current: 0, longest: 0, last_workout_date: null, alive: false, at_risk: false }
  }

  const maxApart = restGap + 1

  let longest = 1
  let chain = 1
  for (let i = 1; i < days.length; i++) {
    chain = dayDiff(days[i - 1], days[i]) <= maxApart ? chain + 1 : 1
    longest = Math.max(longest, chain)
  }

  let currentChain = 1
  let i = days.length - 1
  while (i > 0 && dayDiff(days[i - 1], days[i]) <= maxApart) {
    currentChain += 1
    i -= 1
  }

  const last = days[days.length - 1]
  const elapsed = dayDiff(last, t)
  const alive = elapsed <= maxApart
  return {
    current: alive ? currentChain : 0,
    longest,
    last_workout_date: last,
    alive,
    at_risk: alive && elapsed === maxApart,
  }
}

/** Current streak snapshot — shared by GET /api/streak and the report/dashboard. */
export async function streakSnapshot() {
  const settings = await db.settings.get(1)
  const restGap = settings?.streak_rest_gap ?? DEFAULT_REST_GAP
  const r = computeStreak(await workoutDates(), restGap)
  return { ...r, rest_gap: restGap }
}

export const streakRoutes = [
  { method: 'GET', pattern: /^\/api\/streak$/, handler: () => streakSnapshot() },
]
